src/wasm-lib/kcl/src/errors.rs

```diff
@@ -147,6 +147,157 @@ impl KclErrorWithOutputs {
             filenames: Default::default(),
         }
     }
+    pub async fn into_miette_report_with_outputs(
+        self,
+        fs: std::sync::Arc<crate::fs::FileManager>,
+    ) -> anyhow::Result<ReportWithOutputs> {
+        let mut source_ranges = self.error.source_ranges();
+
+        // Pop off the first source range to get the filename.
+        let first_source_range = source_ranges
+            .pop()
+            .ok_or_else(|| anyhow::anyhow!("No source ranges found"))?;
+
+        let filename = self
+            .filenames
+            .get(&first_source_range.module_id())
+            .map(|p| p.to_string())
+            .ok_or_else(|| {
+                anyhow::anyhow!(
+                    "Could not find filename for module id: {:?}",
+                    first_source_range.module_id()
+                )
+            })?;
+        let kcl_source = std::str::from_utf8(&fs.read(&filename, Default::default()).await?)?.to_string();
+
+        let mut related = Vec::new();
+        for source_range in source_ranges {
+            let module_id = source_range.module_id();
+            let filename = self
+                .filenames
+                .get(&module_id)
+                .map(|p| p.to_string())
+                .ok_or_else(|| anyhow::anyhow!("Could not find filename for module id: {:?}", module_id))?;
+            let error = self.error.override_source_ranges(vec![source_range]);
+            let report = Report {
+                error,
+                kcl_source: std::str::from_utf8(&fs.read(&filename, Default::default()).await?)?.to_string(),
+                filename,
+            };
+            related.push(report);
+        }
+
+        Ok(ReportWithOutputs {
+            error: self,
+            kcl_source,
+            filename,
+            related,
+        })
+    }
+}
+
+impl IntoDiagnostic for KclErrorWithOutputs {
+    async fn to_lsp_diagnostics(&self, code: &str) -> Vec<Diagnostic> {
+        let message = self.error.get_message();
+        let source_ranges = self.error.source_ranges();
+
+        let mut diagnostics = Vec::new();
+
+        for source_range in &source_ranges {
+            let related_information =
+                if let Some(mut filename) = self.filenames.get(&source_range.module_id()).map(|p| p.to_string()) {
+                    let code = std::fs::read_to_string(&filename).unwrap_or(code.to_string());
```

Comment:
Reading the code of a file for every source range seems wasteful. Do we not keep this code around somewhere? I don't see where this is used too, should it end up in the diagnostic?

Reply:
we did not read that file from the lsp, if we have the file contents its in memory but we'd have to pipe it all the way thru unsure how hard that would be

Reply:
looking

Reply:
if i can get it from memory i can remove the async bullshit and then i dont have to worry about reading tio again

Reply:
no we dont store it... i can add it to the ModulePath, If i add to moduleInfo its too late in the game

Reply:
maybe. i make a new one ModuleSource

Reply:
Yeah, we could make a map from ModulePaths to sources early (I guess populated as we open modules?). Or we could build the ModuleInfo table before executing the module without the repr at first (just the source and add the repr after w execute the module)

Reply:
okay i did that, its cleaner and i got rid of async and all the bullshit